### PR TITLE
ENH: include per-device load time if available

### PR DESCRIPTION
**Pull request by klauer**
## Description
Enables per-device load times above a threshold.
Pairs with https://github.com/pcdshub/happi/pull/236

## Motivation and Context
Devices can take a long time to load, and it should not require much digging on the user's part to find it - and help report it to us.

## How Has This Been Tested?
Running xcspython and looking at its output.

## Where Has This Been Documented?
This PR text.

## Screenshots (if appropriate):

<details>

```code
INFO     load_device - Loading at1l0 [pcdsdevices.attenuator.FeeAtt] ...  SUCCESS!
INFO     load_device - Loading at2l0 [pcdsdevices.attenuator.FEESolidAttenuator] ...  SUCCESS! (2.1 s)
INFO     load_device - Loading at2l0_calc [pcdsdevices.attenuator.AttenuatorCalculator_AT2L0] ...  SUCCESS!
INFO     load_device - Loading bt2l0_pleg_vgc01 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading em1l0 [pcdsdevices.energy_monitor.GEM] ...  SUCCESS!
INFO     load_device - Loading em1l0_gem_vgc10 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading em2l0 [pcdsdevices.energy_monitor.GEM] ...  SUCCESS!
INFO     load_device - Loading em2l0_gem_vgc70 [pcdsdevices.valve.VGC] ...  SUCCESS!
Using production server.
INFO     load_device - Loading hxrsss [pcdsdevices.spectrometer.HXRSpectrometer] ...  SUCCESS! (1.0 s)
INFO     load_device - Loading im1l0 [pcdsdevices.pim.XPIM] ...  SUCCESS!
INFO     load_device - Loading im1l0_xtes_vgc01 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading im2l0 [pcdsdevices.pim.XPIM] ...  SUCCESS!
INFO     load_device - Loading im3l0 [pcdsdevices.pim.PPM] ...  SUCCESS!
INFO     load_device - Loading im4l0 [pcdsdevices.pim.XPIM] ...  SUCCESS!
INFO     load_device - Loading mr1l0_homs [pcdsdevices.mirror.XOffsetMirrorState] ...  SUCCESS!
INFO     load_device - Loading mr1l0_homs_vgc01 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading mr2l0_homs [pcdsdevices.mirror.XOffsetMirrorState] ...  SUCCESS!
INFO     load_device - Loading mr2l0_homs_vgc01 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading mr2l0_homs_vgc_02 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading pc1l0_xtes_vgc01 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading pf1l0 [pcdsdevices.device_types.WaveFrontSensorTarget] ...  SUCCESS!
INFO     load_device - Loading rtdsl0 [pcdsdevices.rtds_ebd.RTDSL0] ...  SUCCESS!
INFO     load_device - Loading sl1l0 [pcdsdevices.slits.PowerSlits] ...  SUCCESS!
INFO     load_device - Loading sp1l0 [pcdsdevices.spectrometer.Kmono] ...  SUCCESS!
INFO     load_device - Loading st1l0_pps [pcdsdevices.valve.PPSStopper] ...  SUCCESS!
INFO     load_device - Loading tv1l0_vgc01 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading tv2l0_vgc01 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading tv2l0_vgc02 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading tv3l0_vfs01 [pcdsdevices.valve.VFS] ...  SUCCESS!
INFO     load_device - Loading mr1l3_homs [pcdsdevices.mirror.XOffsetMirror] ...  SUCCESS!
INFO     load_device - Loading mr1l4_homs [pcdsdevices.mirror.XOffsetMirror] ...  SUCCESS!
INFO     load_device - Loading mr2l3_homs [pcdsdevices.mirror.XOffsetMirror] ...  SUCCESS!
INFO     load_device - Loading hx2_ipm [pcdsdevices.device_types.IPM] ...  SUCCESS!
INFO     load_device - Loading hx2_pim [pcdsdevices.pim.PIM] ...  SUCCESS!
INFO     load_device - Loading hx2_reference_laser [pcdsdevices.inout.InOutRecordPositioner] ...  SUCCESS!
INFO     load_device - Loading hx2_slits [pcdsdevices.device_types.Slits] ...  SUCCESS!
INFO     load_device - Loading hxd_dg2_ipm [pcdsdevices.device_types.IPM] ...  SUCCESS!
WARNING  fill_template - Can not enforce type to match attribute prefix_det as it does not exist on the container.
INFO     load_device - Loading hxd_dg2_pim [pcdsdevices.pim.PIM] ...  SUCCESS!
INFO     load_device - Loading hxd_dg2_slits [pcdsdevices.device_types.Slits] ...  SUCCESS!
INFO     load_device - Loading hxd_dg2_stopper [pcdsdevices.device_types.Stopper] ...  SUCCESS!
INFO     load_device - Loading hxd_dg2_valve [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading hxd_mon_valve_1 [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading hxd_mon_valve_2 [pcdsdevices.device_types.GateValve] ...  SUCCESS!
WARNING  fill_template - Can not enforce type to match attribute prefix_det as it does not exist on the container.
INFO     load_device - Loading mec_yag0 [pcdsdevices.pim.PIMWithFocus] ...  SUCCESS!
WARNING  fill_template - Can not enforce type to match attribute prefix_det as it does not exist on the container.
INFO     load_device - Loading mec_yag1 [pcdsdevices.pim.PIMWithFocus] ...  SUCCESS!
INFO     load_device - Loading mxt_valve_1 [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading mxt_valve_2 [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading sh2 [pcdsdevices.valve.PPSStopper] ...  SUCCESS!
INFO     load_device - Loading tv4l0_vgc01 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading tv5l0_vgc01 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading tv6l0_vgc01 [pcdsdevices.valve.VGC] ...  SUCCESS!
INFO     load_device - Loading um6_ipm [pcdsdevices.device_types.IPM] ...  SUCCESS!
WARNING  fill_template - Can not enforce type to match attribute prefix_det as it does not exist on the container.
INFO     load_device - Loading um6_pim [pcdsdevices.pim.PIMWithLED] ...  SUCCESS!
INFO     load_device - Loading um6_slits [pcdsdevices.device_types.Slits] ...  SUCCESS!
INFO     load_device - Loading um6_stopper [pcdsdevices.device_types.Stopper] ...  SUCCESS!
INFO     load_device - Loading xcs_lodcm [pcdsdevices.device_types.LODCM] ...  SUCCESS! (5.6 s)
INFO     load_device - Loading xcs_reference_laser [pcdsdevices.inout.InOutRecordPositioner] ...  SUCCESS!
INFO     load_device - Loading xpp_lodcm [pcdsdevices.device_types.LODCM] ...  SUCCESS! (13.2 s)
INFO     load_device - Loading xrt_dg3m_pim [pcdsdevices.pim.PIM] ...  SUCCESS!
INFO     load_device - Loading xrt_dg3m_slits [pcdsdevices.device_types.Slits] ...  SUCCESS!
INFO     load_device - Loading xrt_dvd_valve [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading xrt_m2h [pcdsdevices.mirror.PointingMirror] ...  SUCCESS!
INFO     load_device - Loading xrt_mxt_valve [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading hfx_mon_valve_3 [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading hfx_mon_valve_4 [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading sh4 [pcdsdevices.valve.PPSStopper] ...  SUCCESS!
INFO     load_device - Loading xcs_attenuator [pcdsdevices.device_types.Attenuator] ...  SUCCESS!
INFO     load_device - Loading xcs_ccm [pcdsdevices.ccm.CCM] ...  SUCCESS! (0.6 s)
INFO     load_device - Loading xcs_dg3_ipm [pcdsdevices.device_types.IPM] ...  SUCCESS!
WARNING  fill_template - Can not enforce type to match attribute prefix_det as it does not exist on the container.
INFO     load_device - Loading xcs_dg3_pim [pcdsdevices.pim.PIMWithLED] ...  SUCCESS!
INFO     load_device - Loading xcs_dg3_slits [pcdsdevices.device_types.Slits] ...  SUCCESS!
INFO     load_device - Loading xcs_dg3_valve_2 [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading xcs_gon [pcdsdevices.gon.GonWithDetArm] ...  SUCCESS!
INFO     load_device - Loading xcs_lam_valve_1 [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading xcs_lam_valve_2 [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading xcs_lxt_fast [pcdsdevices.epics_motor.DelayNewport] ...  SUCCESS!
INFO     load_device - Loading xcs_mon_ipm [pcdsdevices.device_types.IPM] ...  SUCCESS!
INFO     load_device - Loading xcs_pfls [pcdsdevices.lens.Prefocus] ...  SUCCESS!
INFO     load_device - Loading xcs_pulsepicker [pcdsdevices.device_types.PulsePicker] ...  SUCCESS!
INFO     load_device - Loading xcs_samplestage [pcdsdevices.gon.XYZStage] ...  SUCCESS!
INFO     load_device - Loading xcs_sb1_ipm [pcdsdevices.device_types.IPM] ...  SUCCESS!
WARNING  fill_template - Can not enforce type to match attribute prefix_det as it does not exist on the container.
INFO     load_device - Loading xcs_sb1_pim [pcdsdevices.pim.PIMWithLED] ...  SUCCESS!
INFO     load_device - Loading xcs_sb1_slits [pcdsdevices.device_types.Slits] ...  SUCCESS!
INFO     load_device - Loading xcs_sb1_stopper [pcdsdevices.device_types.Stopper] ...  SUCCESS!
INFO     load_device - Loading xcs_sb1_valve [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading xcs_sb2_downstream_slits [pcdsdevices.device_types.Slits] ...  SUCCESS!
INFO     load_device - Loading xcs_sb2_ipm [pcdsdevices.device_types.IPM] ...  SUCCESS!
WARNING  fill_template - Can not enforce type to match attribute prefix_det as it does not exist on the container.
INFO     load_device - Loading xcs_sb2_pim [pcdsdevices.pim.PIMWithLED] ...  SUCCESS!
INFO     load_device - Loading xcs_sb2_upstream_slits [pcdsdevices.device_types.Slits] ...  SUCCESS!
INFO     load_device - Loading xcs_sb2_vgc_01 [pcdsdevices.device_types.GateValve] ...  SUCCESS!
INFO     load_device - Loading xcs_txt [pcdsdevices.epics_motor.DelayNewport] ...  SUCCESS!
INFO     load_device - Loading xcs_xfls [pcdsdevices.device_types.XFLS] ...  SUCCESS!
SUCCESS  safe_load - Successfully loaded database in 32.12 s
```
</details>

### CODE REVIEW

File: hutch_python/happi.py [hunk 1]
```diff
@@ -62,7 +63,12 @@ def get_happi_objs(db, hutch):
             logger.warning("No items found in database for %s",
                            beamline.upper())
     # Instantiate the devices needed
-    dev_namespace = load_devices(*containers, pprint=False)
+    sig = inspect.signature(load_devices)
+    if "include_load_time" in sig.parameters:
+        kwargs = dict(include_load_time=True, load_time_threshold=0.5)
```
Comment (tangkong) [inline hutch_python/happi.py:68]: 0.5s is the default threshold right?  Though not strictly necessary, I think it's fine to include this here if only as a reminder of how to change the parameter.

File: hutch_python/happi.py [hunk 2]
```diff
@@ -62,7 +63,12 @@ def get_happi_objs(db, hutch):
             logger.warning("No items found in database for %s",
                            beamline.upper())
     # Instantiate the devices needed
-    dev_namespace = load_devices(*containers, pprint=False)
+    sig = inspect.signature(load_devices)
```
Comment (tangkong) [inline hutch_python/happi.py:66]: We do this check in case happi hasn't been updated yet?

File: hutch_python/happi.py [hunk 3]
```diff
@@ -62,7 +63,12 @@ def get_happi_objs(db, hutch):
...
+    if "include_load_time" in sig.parameters:
```
Comment (tangkong) [inline hutch_python/happi.py:67]: We do this check in case Happi hasn't been updated?  Once it has been do we need to inspect `load_devices`?

Comment (klauer) [reply hutch_python/happi.py:67]: In my head it was "what if we have a dev install of happi without the parameter (updated hutch-python but not happi)". Despite being the one to propose it, I think this is overkill, and would be happy to remove it if you two agree.

Comment (ZLLentz) [reply hutch_python/happi.py:67]: Normally this would be overkill but we also have arbitrary dev installs and I've been bitten in the past by this

Comment (ZLLentz) [reply hutch_python/happi.py:67]: Eh, it's probably OK to assume updated happi, we should go tag it soon

Comment (tangkong) [reply hutch_python/happi.py:67]: I'm all for being safe, I guess the question becomes more of what extent to be backward compatible.  I'm fine leaving it as well, fwiw

**Review by ZLLentz — APPROVED** [review]

File: hutch_python/happi.py (hunk 3 above)
Comment (klauer) [reply hutch_python/happi.py:67]: Future selves: we decided that the lazy answer was the right one, but we can come back and nuke this down the line.

**Pull request merged**